src/persist/src/postgres.rs

```diff
@@ -339,39 +340,70 @@ impl Consensus for PostgresConsensus {
             }
         }
 
-        let result = if let Some(expected) = expected {
-            // Only insert the new row if:
-            // - sequence number expected is already present
-            // - expected corresponds to the most recent sequence number
-            //   i.e. there is no other sequence number > expected already
-            //   present.
-            //
-            // This query has also been written to execute within a single
-            // network round-trip (instead of a slightly simpler implementation
-            // that would call `BEGIN` and have multiple `SELECT` queries).
-            let q = "INSERT INTO consensus SELECT $1, $2, $3 WHERE
-                     EXISTS (
-                        SELECT * FROM consensus WHERE shard = $1 AND sequence_number = $4
-                     )
-                     AND NOT EXISTS (
-                         SELECT * FROM consensus WHERE shard = $1 AND sequence_number > $4
-                     )
-                     ON CONFLICT DO NOTHING";
-            let client = self.get_connection().await?;
-            let statement = client.prepare_cached(q).await?;
-            client
-                .execute(
-                    &statement,
-                    &[&key, &new.seqno, &new.data.as_ref(), &expected],
+        if let Some(expected) = expected {
+            // This query is tuned for performance against CockroacbDB.
+            // The read for the latest sequence number should hit 1 row
+            // via reverse scan over the PK, and the insert is designed
+            // to hit the fast-path 1-phase commit of CRDB. Any changes
+            // to this query should ensure a  EXPLAIN ANALYZE (VERBOSE)
+            // query plan contains `auto commit` on the insert.
+            let cas_insert_query = r#"
+                WITH
+                    greatest_seqno AS (
+                        SELECT sequence_number FROM consensus
+                        WHERE shard = $1
+                        ORDER BY sequence_number DESC LIMIT 1
+                    )
+                    INSERT INTO consensus (shard, sequence_number, data)
+                    SELECT $1, $2, $3
+                    WHERE (SELECT * FROM greatest_seqno) = $4;
+            "#;
+            // We pipeline an additional read query to fetch the latest state.
```

Comment:
clever! were you able to confirm that this actually pipelines in the way we think (perhaps by looking at latencies with and without it)?